[WIP] Qt6 - Bringing JASP up to speed

JorisGoosen opened the pull request:
Yeah so they changed a lot of stuff and I wanted to test something with the new Qt.

So this is a corrolary to https://github.com/jasp-stats/jasp-desktop/pull/4658 in such a way that @amirmasoudabdol can get notifications about it.

amirmasoudabdol committed:
Added the boost:: namespace to all the filesystem. This resolves a conflict when we move to the C++17

amirmasoudabdol committed:
replaced the deprecated QSet::toList()

amirmasoudabdol committed:
Replaced the deprecated `QString::split(QRegEx ...)` with `QString::split(QRegularExpression, ...)`

amirmasoudabdol committed:
Resolved the casting error. I don't know why we need to do get it to QChar, but if we have to, we can do `QChar point = *locale().decimalPoint().toStdString().c_str();`

amirmasoudabdol committed:
This will cause some confusion for the compiler later when the mocing happens. It's a `bool` anyway!

amirmasoudabdol committed:
- Added the missing const operator

amirmasoudabdol committed:
These are not ideal, but C++17 will complain if the types do not match, because `std::max`, `std::min` are template functions. So, we need to consider this. There are more instance of them around which we need to deal with. They are not causing error because they are safer casts but still.

amirmasoudabdol committed:
Replaced the qSort with std::sort. qSort is deprecated, and std::sort is N * Log(N) as well.

amirmasoudabdol committed:
Replaced `qsrand` and `qrand` with `std::srand`, and `std::rand`. Those two are discontinued. If we care, we can do better in c++11/17 regarding the random number

amirmasoudabdol committed:
Using QIODeviceBase instead of `QIODevice` for getting the `OpenModeFlag`. They are not in the `QIODevice` anymore!

amirmasoudabdol committed:
Replaced `midRef` and `leftRef` with `QtString::mid` and `QtString::left`. They are the same and old ones are gone! Btw, they return QString anyway!

amirmasoudabdol committed:
The old one is obsolete

amirmasoudabdol committed:
I'm not so sure about this one, but I think that's you wanted to do. From 6.1, we can use `QLocal::codeToLanguage` to deal with this. We can perhaps replace the entire `isValidLocal` with it since it returns a proper error

amirmasoudabdol committed:
`QDateTime::toTime_t()` is obsolete.

amirmasoudabdol committed:
:exclamation: This is probably not exactly what we want, but getting it right was beyond the focus of this branch. I'm mainly trying to get the 6.2 to work at this point. I'm not even sure what this file does!

amirmasoudabdol committed:
This is an infinite loop. `QStringList.toVector()` spits out `QList<QString>`. Qt Creator picks it up, and shows a warning for it.

I really hope there were a line on top of these file describing their purpose. I have no idea what this file does, or what `fq`, `tq, `fqj` mean!

amirmasoudabdol committed:
Some more minor adaptation. Overall, c++17 is even more strict with types because of the templates and enum classes and Qt 6.2 follows along which is very nice in fact.

- The `UTF-8` is the default option
- The `GL_Lines` is now inside an enum class
- The `fromStdVector` is deprecated. They are relying on `initializer_list` overall which is very nice too. I used a new syntax there, but we will see if we like it or not
- and finally inserted some missing headers

amirmasoudabdol committed:
By the last commit, at least C++ doesn't nag anymore. So, here we are activating the c++17 flag, and moving to the QML side of things. I had to make a few changes to the QMake commands to get a proper c++17 build, but they can be removed or simplified as we move to CMake. I also activated some of the errors to make sure that at this stage I get most of the tricky stuff, of course we can disable them later.

amirmasoudabdol committed:
:warning: Disabled the shadow effects, by removing the `QtGraphicalEffects`. This has moved into the `Qt5Compat.GraphicalEffects` and I expect it to either be implemented somewhere else, or be deprecated soon. We can bring it back if we need to.

amirmasoudabdol committed:
:warning: Replaced the instances of QtQuick.Controls 1.4 with QtQuick.Controls 6.0 wherever was needed! I think Qt is changing their naming scheme, and they are giving their control an umbrella version number which is very nice, but that might cause some issues for us. I think this breaks the SplitView but I cannot immediately see why. I'm marking these commits that we can go back to them, and evaluate our choices

amirmasoudabdol committed:
:exclamation: These are breaking changes, and we need to find a workaround for them

- Styling has changed and this is not valid anymore. I think this might be breaking the AboutWindows but I'm not sure. I feel that's related to the WebEngineQuick. See next commit.
- Commented a bug regarding the shortcut, this should be done differently but I am leaving it for later. I have a list! :P
- I don't think Button has `checkable` and `checked`, maybe it has but Qt doesn't like it. It's not also our Button. I checked!

amirmasoudabdol committed:
:exclamation: Successful build, but broken app!

- Replacing the `QtWebEngine` with `QtWebEngineQuick`. This is explained [here](https://doc.qt.io/qt-6/qtwebengine-changes-qt6.html). One thing that is strange to me is the initialization of the WebEngine. I see Joris' comment, but then [here](https://doc.qt.io/qt-6/qtwebenginequick.html#initialize) it says that it should be initialized before the instantiation of the QApplication.
- I traced back a crash to the `DataSetPackage::index` where that point was causing some termination. I am not 100% sure what's happening there, but I'm not the fan of those pointers. I have tried to deal with it in different ways, but I didn't manage to get it to behave. I think there is something with the enum that breaks it, and leaving it uninitialized at least gives me the build.

JorisGoosen committed:
Got it to "compile" and "start"
It doesnt work though

JorisGoosen committed:
Fixing some Qml

JorisGoosen committed:
some comments on qutils.

Also because QVector is now a subclass/copy of QList certain code wasnt necessary anymore as Amir found out already when he removed:
`inline	std::vector<std::string>			fq (const QStringList						 & vec)		{ return fq(vec.toVector()); }`

JorisGoosen committed:
Hmm, this is broken!

but a step in the right direction nonetheless.
The pointer structure of datasetpackage might need to be reconsidered as there are some strange infinite loops going on here.

JorisGoosen committed:
Moving to a more sensible structure but CDB is not helping me out by getting stuck reading values from running JASP

I guess I need to boot to linux to get some work done

I guess I should really make it a tree structure

amirmasoudabdol commented on Desktop/data/datasetpackage.cpp:
```diff
@@ -205,22 +267,26 @@ int DataSetPackage::rowCount(const QModelIndex & parent) const
-	case parIdxType::filter:
-	case parIdxType::root:		//return int(parIdxType::leaf); Its more logical to get the actual datasize
-	case parIdxType::data:		return !_dataSet ? 0 : _dataSet->rowCount();
+	case parIdxType::filterRoot:	[[clang::fallthrough]];
```
I think you can just use `[[fallthrough]]`, no?

JorisGoosen replied:
I can definitely try, that would look a lot better.

JorisGoosen replied:
Well it does indeed work, maybe the last time I tried that we hadn't switched to >C++11 and it was missing?
Anyway, much better like this.

amirmasoudabdol replied:
That could very well be. Full C++17 is not very new, especially on GCC.

JorisGoosen replied:
Haha, no I am aware we are living in the future, but it took us a while to switch to >11 even.

JorisGoosen committed:
more attempts but I should just continue tomorrow and rebase+squash this away

JorisGoosen committed:
added some submodels to pre-filter datasetpackage before handing it over to the sortfilterproxymodels (which have internal mapping so also filtering rootnodes is impractical)

JorisGoosen committed:
maybe using a modified QIdentityProxyModel is better than QAbstractProxyModel as the latter doesnt seem to connect signals and so on.

JorisGoosen committed:
alright data is finally being passed through and JASP doesnt seem to crash.
Except if I turn on the nice custom rendering lines in DataSetView...
So I've disabled them for now.

JorisGoosen committed:
label can now be selected

JorisGoosen committed:
Alright, the values and labels are gotten properly now and some more qml warnings silenced

JorisGoosen committed:
Enabling and fixing checked and checkable in VariablesWindow

JorisGoosen committed:
almost got the splitviews to behave again as well

JorisGoosen committed:
something weird is going on with the splithandlers

JorisGoosen committed:
Results and data panel are now accessible and visible

JorisGoosen committed:
More functionparameters for onSignal handlers

JorisGoosen committed:
make sure IPCChannel is always created *before* starting JASPEngine

JorisGoosen committed:
Implements EJs request for moving forms a bit out of screen lefthand if wanted

Make EJs desire work better

JorisGoosen committed:
- Remove old unused file BasicTableViewJASP
- make sure suggested column icons in variableslists actually change with the theme.
- undo some breaking changes by Amir

JorisGoosen committed:
Put shadows back

JorisGoosen committed:
Apparently the lines in DataSetView work again so I can enable them then

JorisGoosen committed:
Convert shaders to separate files so they can later be converted into .qsb
This can be easily done through cmake as described here: https://doc.qt.io/qt-6/qtshadertools-build.html

Rewrite shaders to work with qsb

for now, until cmake is done, I compiled them as:

D:\Broncode\JASP\jasp-desktop\Desktop\components\JASP\Widgets>D:\Qt\6.2.1\msvc2019_64\bin\qsb  --glsl 100es,120,150 --hlsl 50 --msl 12 -b -o ImageInverter.vert.qsb ImageInverter.vert
D:\Broncode\JASP\jasp-desktop\Desktop\components\JASP\Widgets>D:\Qt\6.2.1\msvc2019_64\bin\qsb  --glsl 100es,120,150 --hlsl 50 --msl 12 -o ImageInverter.frag.qsb ImageInverter.frag

JorisGoosen committed:
Fixing some qml (arguments for handlers)

JorisGoosen committed:
Make special plot-scheme-handler to avoid annoying "not allowed to load local resources" problem with Qt6

JorisGoosen committed:
Fixes and refactoring

- no need for singleShotTime with queued connection
- also some changes in Qt6 broke `source: "repeatedMeasures"` as opposed to `source: ["repeatedMeasures"]`, ive added a workaround.

JorisGoosen committed:
Make sure data is hidden if a jasp file including analyses is loaded

JorisGoosen committed:
AnalysisFormExpander now opens smoothly again

JorisGoosen committed:
new major R-Interface number

amirmasoudabdol commented:
I'm testing this a bit, and it looks mostly fine, but I see this strange/bad behavior with the help form, and resizing. I guess this only happens on Mac but we need to fix it. There are cases where you cannot see anything on JASP, help, or both windows. There are some random crashes that I cannot reproduce as well. Sometimes the Engine just dies which is weird!

![2022-01-07 11 39 47](https://user-images.githubusercontent.com/1290841/148532443-c2a6de9a-37e9-4c7c-8bc4-8c319e89ec04.gif)

<img width="1316" alt="CleanShot 2022-01-07 at 11 26 20@2x" src="https://user-images.githubusercontent.com/1290841/148532588-dfb2fd2a-2a2d-4f1e-b59c-2918f23a652d.png">

I talked to Tim and Bruno, and I think we can merge this right after releasing the 0.16.1, and provide nightlies for those who are facing the slowness issues. In the meantime, we can refine and test it and prepare it for 0.17. However, if we see it being stable during the testing, we can have a small release and only push this patch out as 0.16.2.

JorisGoosen commented:
Pfooh, the flickering looks bad!

Did this also occur on normal mac? 

If so I think your suggestion is a good one. But I would still want to put it in 0.16.1 for windows (or at least a quick release) because this bug is now forcing people to work with 0.14 or something ancient like that...

AlexanderLyNL commented:
I'd prefer to have 0.16.1 out without this. Once that's out we can upgrade to Qt 6 so the upgrade is isolated

JorisGoosen commented:
> I'd prefer to have 0.16.1 out without this. Once that's out we can upgrade to Qt 6 so the upgrade is isolated

Ok but what about https://github.com/jasp-stats/jasp-issues/issues/1490 https://github.com/jasp-stats/jasp-issues/issues/1431 ?

Also, at least 1 person at SKF has this same problem. 
I do agree that it is nice to deliver the same version to all systems but also: having a usable JASP on all windows machines would be a good thing right? Most users are on windows after all.

EJWagenmakers commented:
A compromise could be to do 0.16.1 without the upgrade, but a quick 0.16.2 solely with that upgrade. More work though.

JorisGoosen commented:
> A compromise could be to do 0.16.1 without the upgrade, but a quick 0.16.2 solely with that upgrade. More work though.

More work, but also, the others are right that the changes need to be tested and this way we achieve that very nicely. Also gives users a relatively light way to still use the new stuff in 0.16.1 without Qt6 and the accompanying changes messing it up.

amirmasoudabdol commented:
I think Qt 6.2.3 has fixed the flickering issue on macOS! 👏🏼

boutinb commented:
> I think Qt 6.2.3 has fixed the flickering issue on macOS! 👏🏼

That's another really good news!

JorisGoosen committed:
Up minor version, it is now at 0.16.2
